### ISSUE TITLE
Is it possible to have too many asserts?

### ISSUE THREAD

**Issue by NorbertLoh** [nus-cs2103-AY2425S1/forum#598]
We are told to use asserts liberally. However for example, if we have a constructor with 3 parameters, would writing asserts for all boundary values for all parameters be too much? This could easily lead to at least 10 assert statements.

Perhaps this can be used as an example.
![image](https://github.com/user-attachments/assets/810f739d-1260-4ee4-9ed8-52485aa29d37)

**Comment by Incogdino** [comment]
I do not see a reason why there should be a cap to the limit on assertions. 

But it seems as though your assertions are calling other methods here. Without knowing too much details of what the methods does, I do not think is very much a good idea. 

For instance, `model.hasLessons(lesson)` should be more like an exception (or a guard clause) rather than an assertion.

Assertions are used to indicate when the programmer screws up and exceptions for when the user or program screws up. In this case, it seem more likely that the assertion happens when the program screws up (bug in code) and should be handled by an exception instead.

**Comment by Incogdino** [comment]
Also, when the program runs without assertions,  and the bug happens to go through (without the assertion) then the program will just crash/ not work as expected

**Comment by NorbertLoh** [comment]
@Incogdino I am calling other methods for the assertions that are calling them to check that the post conditions are correct so that if the original guarding method was changed incorrectly, the assertion would pick it up. 
However, using too many asserts makes the code look quite long and untidy, although without much performance isues. Therefore, I was wondering where to draw the line.

**Comment by NgZiXin** [comment]
I feel that some of the asserts may not be necessary here because other classes like model has its own defensive checks. E.g. If you try to add a duplicated person to the model, the UniquePersonList will throw an exception.

**Comment by yadobler** [comment]
**TLDR**

- Performance wise: no impact (on release). 
- Use until ~~shiok~~ you feel it is enough / it starts diminishing in value of readability
- asserts are a _debugging / development_ tool - assume they will vanish in prod and use guard clauses instead for unavoidable exception states that can happen in prod, so that the programme _gently_ errors out and not explode in the user's face or silently run in an invalid state. 

----

Java is mostly JIT-compilation and without explicit `-ea` flag, the JIT-compiler will drop the `assert` bytecodes. `assert` statements are in theory supposed to be ignored in release files (ie for debugging purposes). You cannot have assert statements annihilating the whole program - you want to fail silently (and *log* it). So in testing, you'd run `java` with `-ea` flag (**enable assertions**). 

By default, the behaviour is to run with assertions disabled (`-da`)[^1]. 

[^1]: More on the flag [here](https://docs.oracle.com/cd/E19683-01/806-7930/6jgp65ikq/index.html), including how to specifically choose which class to enable and disable assertions with the cli flags

----

2103 being 2103, it's a perfect qn to ask in terms of code quality, ask yourself:

**Can you understand the code you wrote yesterday with the assert statements everywhere?**:
- Yes: ~~not enough asserts~~ it's fine
- No: stop. 
  - That being said, I feel like asserts should ideally start/end a segment of code. You need to break your wall of code with newlines. Else it feels like reading an essay with no paragraphs. 
  - `assert model != null` is duplicating the `requireNonNull(model)` code. 

----

I also agree with @Incogdino. Ask yourself if the checks are:
- recoverable exceptions; or
- unrecoverable errors / invalid states

A user entering invalid command is still a valid state with an exception that you raise and then catch and process. That should be *in the release*. 

A user entering invalid command and the app successfully accepts without throwing exception? That is an invalid state - it should be caught in development and the program must burn and crash. 

If you think it that a "not happy" state can be encountered in production and should be *recoverable*, hmhm, like how AB3 is able to populate a new `addressbook.json` if it is missing / is corrupted (This means it is not an _invalid state_ - it's just another valid state with exception), then you may want to "incorporate" this assert checking into actual code as an exception instead.

**Comment by NorbertLoh** [comment]
> I feel that some of the asserts may not be necessary here because other classes like model has its own defensive checks. E.g. If you try to add a duplicated person to the model, the UniquePersonList will throw an exception.

I agree. Most of our methods already have their own defensive checks. So I am unsure as to when an assert statement will actually be helpful.
